PC/_wmimodule.cpp

```diff
@@ -0,0 +1,310 @@
+
+Runs a WMI query against the local machine.
+
+This returns a single string with 'name=value' pairs in a flat array separated
+by null characters.
+[clinic start generated code]*/
+
+static PyObject *
+_wmi_exec_query_impl(PyObject *module, PyObject *query)
+/*[clinic end generated code: output=a62303d5bb5e003f input=48d2d0a1e1a7e3c2]*/
+
+/*[clinic end generated code]*/
+{
+    PyObject *result = NULL;
+    HANDLE hThread = NULL;
+    int err = 0;
+    WCHAR buffer[8192];
+    DWORD offset = 0;
+    DWORD bytesRead;
+    struct _query_data data = {0};
+
+    if (PySys_Audit("_wmi.exec_query", "O", query) < 0) {
+        return NULL;
+    }
+
+    data.query = PyUnicode_AsWideCharString(query, NULL);
+    if (!data.query) {
+        return NULL;
+    }
+
+    if (0 != _wcsnicmp(data.query, L"select ", 7)) {
+        PyMem_Free((void *)data.query);
+        PyErr_SetString(PyExc_ValueError, "only SELECT queries are supported");
+        return NULL;
+    }
+
+    Py_BEGIN_ALLOW_THREADS
+
+    if (!CreatePipe(&data.readPipe, &data.writePipe, NULL, 0)) {
+        err = GetLastError();
+    } else {
+        hThread = CreateThread(NULL, 0, _query_thread, (LPVOID*)&data, 0, NULL);
+        if (!hThread) {
+            err = GetLastError();
+            // Normally the thread proc closes this handle, but since we never started
+            // we need to close it here.
+            CloseHandle(data.writePipe);
+        }
+    }
+
+    while (!err) {
+        if (ReadFile(
+            data.readPipe,
+            (LPVOID)&buffer[offset / sizeof(buffer[0])],
+            sizeof(buffer) - offset,
+            &bytesRead,
+            NULL
+        )) {
+            offset += bytesRead;
+            if (offset >= sizeof(buffer)) {
+                err = ERROR_MORE_DATA;
+            }
+        } else {
+            err = GetLastError();
+        }
+    }
+
+    if (data.readPipe) {
+        CloseHandle(data.readPipe);
+    }
+
+    // Allow the thread some time to clean up
+    switch (WaitForSingleObject(hThread, 1000)) {
+    case WAIT_OBJECT_0:
+        // Thread ended cleanly
+        if (!GetExitCodeThread(hThread, (LPDWORD)&err)) {
+            err = GetLastError();
+        }
+        break;
+    case WAIT_TIMEOUT:
+        // Probably stuck - there's not much we can do, unfortunately
+        if (err == 0 || err == ERROR_BROKEN_PIPE) {
+            err = WAIT_TIMEOUT;
+        }
+        break;
+    default:
+        if (err == 0 || err == ERROR_BROKEN_PIPE) {
+            err = GetLastError();
+        }
+        break;
+    }
+
+    CloseHandle(hThread);
+    hThread = NULL;
+
+    Py_END_ALLOW_THREADS
+
+    PyMem_Free((void *)data.query);
+
+    if (err == ERROR_MORE_DATA) {
+        PyErr_SetString(PyExc_OSError, "Query returns too much data");
+        return NULL;
+    } else if (err) {
+        PyErr_SetFromWindowsErr(err);
+        return NULL;
+    }
+
+    return PyUnicode_FromWideChar(buffer, offset  / sizeof(buffer[0]) - 1);
```

Comment:
What if offset is 0? Maybe just return an empty string in that case? Something like:

```suggestion
    Py_ssize_t len = offset / sizeof(buffer[0]);
    if (len != 0) {
        len--;
    }
    return PyUnicode_FromWideChar(buffer, len);
```

or:

```
if (offset < 1) return PyUnicode_FromStringAndSize(NULL, 0);
```

Reply:
```suggestion
    if (!offset) {
        return PyUnicode_FromStringAndSize(NULL, 0);
    }
    return PyUnicode_FromWideChar(buffer, offset  / sizeof(buffer[0]) - 1);
```